File: 09-background-generator/LightXBackgroundGenerator.kt
/**
 * LightX AI Background Generator API Integration - Kotlin
 * 
 * This implementation provides a complete integration with LightX API v2
 * for AI-powered background generation functionality.
 */

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.io.*
import java.net.*
import java.net.http.*
import java.time.Duration
import java.awt.image.BufferedImage
import javax.imageio.ImageIO

// MARK: - Data Classes

@Serializable
data class BackgroundGeneratorUploadImageResponse(
    val statusCode: Int,
    val message: String,
    val body: BackgroundGeneratorUploadImageBody
)

@Serializable
data class BackgroundGeneratorUploadImageBody(
    val uploadImage: String,
    val imageUrl: String,
    val size: Int
)

@Serializable
data class BackgroundGeneratorGenerationResponse(
    val statusCode: Int,
    val message: String,
    val body: BackgroundGeneratorGenerationBody
)

@Serializable
data class BackgroundGeneratorGenerationBody(
    val orderId: String,
    val maxRetriesAllowed: Int,
    val avgResponseTimeInSec: Int,
    val status: String
)

@Serializable
data class BackgroundGeneratorOrderStatusResponse(
    val statusCode: Int,
    val message: String,
    val body: BackgroundGeneratorOrderStatusBody
)

@Serializable
data class BackgroundGeneratorOrderStatusBody(
    val orderId: String,
    val status: String,
    val output: String? = null
)

// MARK: - LightX Background Generator API Client

class LightXBackgroundGeneratorAPI(private val apiKey: String) {
    
    companion object {
        private const val BASE_URL = "https://api.lightxeditor.com/external/api"
        private const val MAX_RETRIES = 5
        private const val RETRY_INTERVAL = 3000L // milliseconds
        private const val MAX_FILE_SIZE = 5242880L // 5MB
    }
    
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(30))
        .build()
    
    private val json = Json { ignoreUnknownKeys = true }
    
    /**
     * Upload image to LightX servers
     * @param imageFile File containing the image data
     * @param contentType MIME type (image/jpeg or image/png)
     * @return Final image URL
     */
    suspend fun uploadImage(imageFile: File, contentType: String = "image/jpeg"): String {
        val fileSize = imageFile.length()
        
        if (fileSize > MAX_FILE_SIZE) {
            throw LightXBackgroundGeneratorException("Image size exceeds 5MB limit")
        }
        
        // Step 1: Get upload URL
        val uploadUrl = getUploadURL(fileSize, contentType)
        
        // Step 2: Upload image to S3
        uploadToS3(uploadUrl.uploadImage, imageFile, contentType)
        
        println("✅ Image uploaded successfully")
        return uploadUrl.imageUrl
    }
    
    /**
     * Generate background
     * @param imageUrl URL of the input image
     * @param textPrompt Text prompt for background generation
     * @return Order ID for tracking
     */
    suspend fun generateBackground(imageUrl: String, textPrompt: String): String {
        val endpoint = "$BASE_URL/v1/background-generator"
        
        val requestBody = buildJsonObject {
            put("imageUrl", imageUrl)
            put("textPrompt", textPrompt)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXBackgroundGeneratorException("Network error: ${response.statusCode()}")
        }
        
        val backgroundResponse = json.decodeFromString<BackgroundGeneratorGenerationResponse>(response.body())
        
        if (backgroundResponse.statusCode != 2000) {
            throw LightXBackgroundGeneratorException("Background generation request failed: ${backgroundResponse.message}")
        }
        
        val orderInfo = backgroundResponse.body
        
        println("📋 Order created: ${orderInfo.orderId}")
        println("🔄 Max retries allowed: ${orderInfo.maxRetriesAllowed}")
        println("⏱️  Average response time: ${orderInfo.avgResponseTimeInSec} seconds")
        println("📊 Status: ${orderInfo.status}")
        println("💬 Text prompt: \"$textPrompt\"")
        
        return orderInfo.orderId
    }
    
    /**
     * Check order status
     * @param orderId Order ID to check
     * @return Order status and results
     */
    suspend fun checkOrderStatus(orderId: String): BackgroundGeneratorOrderStatusBody {
        val endpoint = "$BASE_URL/v1/order-status"
        
        val requestBody = buildJsonObject {
            put("orderId", orderId)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXBackgroundGeneratorException("Network error: ${response.statusCode()}")
        }
        
        val statusResponse = json.decodeFromString<BackgroundGeneratorOrderStatusResponse>(response.body())
        
        if (statusResponse.statusCode != 2000) {
            throw LightXBackgroundGeneratorException("Status check failed: ${statusResponse.message}")
        }
        
        return statusResponse.body
    }
    
    /**
     * Wait for order completion with automatic retries
     * @param orderId Order ID to monitor
     * @return Final result with output URL
     */
    suspend fun waitForCompletion(orderId: String): BackgroundGeneratorOrderStatusBody {
        var attempts = 0
        
        while (attempts < MAX_RETRIES) {
            try {
                val status = checkOrderStatus(orderId)
                
                println("🔄 Attempt ${attempts + 1}: Status - ${status.status}")
                
                when (status.status) {
                    "active" -> {
                        println("✅ Background generation completed successfully!")
                        status.output?.let { println("🖼️ Background image: $it") }
                        return status
                    }
                    "failed" -> throw LightXBackgroundGeneratorException("Background generation failed")
                    "init" -> {
                        attempts++
                        if (attempts < MAX_RETRIES) {
                            println("⏳ Waiting ${RETRY_INTERVAL / 1000} seconds before next check...")
                            delay(RETRY_INTERVAL)
                        }
                    }
                    else -> {
                        attempts++
                        if (attempts < MAX_RETRIES) {
                            delay(RETRY_INTERVAL)
                        }
                    }
                }
                
            } catch (e: Exception) {
                attempts++
                if (attempts >= MAX_RETRIES) {
                    throw e
                }
                println("⚠️  Error on attempt $attempts, retrying...")
                delay(RETRY_INTERVAL)
            }
        }
        
        throw LightXBackgroundGeneratorException("Maximum retry attempts reached")
    }
    
    /**
     * Complete workflow: Upload image and generate background
     * @param imageFile Image file
     * @param textPrompt Text prompt for background generation
     * @param contentType MIME type
     * @return Final result with output URL
     */
    suspend fun processBackgroundGeneration(
        imageFile: File, 
        textPrompt: String, 
        contentType: String = "image/jpeg"
    ): BackgroundGeneratorOrderStatusBody {
        println("🚀 Starting LightX AI Background Generator API workflow...")
        
        // Step 1: Upload image
        println("📤 Uploading image...")
        val imageUrl = uploadImage(imageFile, contentType)
        println("✅ Image uploaded: $imageUrl")
        
        // Step 2: Generate background
        println("🖼️ Generating background...")
        val orderId = generateBackground(imageUrl, textPrompt)
        
        // Step 3: Wait for completion
        println("⏳ Waiting for processing to complete...")
        val result = waitForCompletion(orderId)
        
        return result
    }
    
    /**
     * Get background generation tips and best practices
     * @return Map of tips for better background results
     */
    fun getBackgroundTips(): Map<String, List<String>> {
        val tips = mapOf(
            "input_image" to listOf(
                "Use clear, well-lit photos with good contrast",
                "Ensure the subject is clearly visible and centered",
                "Avoid cluttered or busy backgrounds",
                "Use high-resolution images for better results",
                "Good subject separation improves background generation"
            ),
            "text_prompts" to listOf(
                "Be specific about the background style you want",
                "Mention color schemes and mood preferences",
                "Include environmental details (indoor, outdoor, studio)",
                "Specify lighting preferences (natural, dramatic, soft)",
                "Keep prompts concise but descriptive"
            ),
            "general" to listOf(
                "Background generation works best with clear subjects",
                "Results may vary based on input image quality",
                "Text prompts guide the background generation process",
                "Allow 15-30 seconds for processing",
                "Experiment with different prompt styles for variety"
            )
        )
        
        println("💡 Background Generation Tips:")
        for ((category, tipList) in tips) {
            println("$category: $tipList")
        }
        return tips
    }
    
    /**
     * Get background style suggestions
     * @return Map of background style suggestions
     */
    fun getBackgroundStyleSuggestions(): Map<String, List<String>> {
        val styleSuggestions = mapOf(
            "professional" to listOf(
                "professional office background",
                "clean minimalist workspace",
                "modern corporate environment",
                "elegant business setting",
                "sophisticated professional space"
            ),
            "natural" to listOf(
                "natural outdoor landscape",
                "beautiful garden background",
                "scenic mountain view",
                "peaceful forest setting",
                "serene beach environment"
            ),
            "creative" to listOf(
                "artistic studio background",
                "creative workspace environment",
                "colorful abstract background",
                "modern art gallery setting",
                "inspiring creative space"
            ),
            "lifestyle" to listOf(
                "cozy home interior",
                "modern living room",
                "stylish bedroom setting",
                "contemporary kitchen",
                "elegant dining room"
            ),
            "outdoor" to listOf(
                "urban cityscape background",
                "park environment",
                "outdoor cafe setting",
                "street scene background",
                "architectural landmark view"
            )
        )
        
        println("💡 Background Style Suggestions:")
        for ((category, suggestionList) in styleSuggestions) {
            println("$category: $suggestionList")
        }
        return styleSuggestions
    }
    
    /**
     * Get background prompt examples
     * @return Map of prompt examples
     */
    fun getBackgroundPromptExamples(): Map<String, List<String>> {
        val promptExamples = mapOf(
            "professional" to listOf(
                "Modern office with glass windows and city view",
                "Clean white studio background with soft lighting",
                "Professional conference room with wooden furniture",
                "Contemporary workspace with minimalist design",
                "Elegant business environment with neutral colors"
            ),
            "natural" to listOf(
                "Beautiful sunset over mountains in the background",
                "Lush green garden with blooming flowers",
                "Peaceful lake with reflection of trees",
                "Golden hour lighting in a forest setting",
                "Serene beach with gentle waves and palm trees"
            ),
            "creative" to listOf(
                "Artistic studio with colorful paint splashes",
                "Modern gallery with white walls and track lighting",
                "Creative workspace with vintage furniture",
                "Abstract colorful background with geometric shapes",
                "Bohemian style room with eclectic decorations"
            ),
            "lifestyle" to listOf(
                "Cozy living room with warm lighting and books",
                "Modern kitchen with marble countertops",
                "Stylish bedroom with soft natural light",
                "Contemporary dining room with elegant table",
                "Comfortable home office with plants and books"
            ),
            "outdoor" to listOf(
                "Urban cityscape with modern skyscrapers",
                "Charming street with cafes and shops",
                "Park setting with trees and walking paths",
                "Historic architecture with classical columns",
                "Modern outdoor space with contemporary design"
            )
        )
        
        println("💡 Background Prompt Examples:")
        for ((category, exampleList) in promptExamples) {
            println("$category: $exampleList")
        }
        return promptExamples
    }
    
    /**
     * Validate text prompt (utility function)
     * @param textPrompt Text prompt to validate
     * @return Whether the prompt is valid
     */
    fun validateTextPrompt(textPrompt: String): Boolean {
        if (textPrompt.trim().isEmpty()) {
            println("❌ Text prompt cannot be empty")
            return false
        }
        
        if (textPrompt.length > 500) {
            println("❌ Text prompt is too long (max 500 characters)")
            return false
        }
        
        println("✅ Text prompt is valid")
        return true
    }
    
    /**
     * Generate background with prompt validation
     * @param imageFile Image file
     * @param textPrompt Text prompt for background generation
     * @param contentType MIME type
     * @return Final result with output URL
     */
    suspend fun generateBackgroundWithPrompt(
        imageFile: File, 
        textPrompt: String, 
        contentType: String = "image/jpeg"
    ): BackgroundGeneratorOrderStatusBody {
        if (!validateTextPrompt(textPrompt)) {
            throw LightXBackgroundGeneratorException("Invalid text prompt")
        }
        
        return processBackgroundGeneration(imageFile, textPrompt, contentType)
    }
    
    /**
     * Get image dimensions (utility function)
     * @param imageFile File containing the image data
     * @return Pair of (width, height)
     */
    fun getImageDimensions(imageFile: File): Pair<Int, Int> {
        return try {
            val image = ImageIO.read(imageFile)
            val width = image.width
            val height = image.height
            println("📏 Image dimensions: ${width}x${height}")
            Pair(width, height)
        } catch (e: Exception) {
            println("❌ Error getting image dimensions: ${e.message}")
            Pair(0, 0)
        }
    }
    
    // MARK: - Private Methods
    
    private suspend fun getUploadURL(fileSize: Long, contentType: String): BackgroundGeneratorUploadImageBody {
        val endpoint = "$BASE_URL/v2/uploadImageUrl"
        
        val requestBody = buildJsonObject {
            put("uploadType", "imageUrl")
            put("size", fileSize)
            put("contentType", contentType)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXBackgroundGeneratorException("Network error: ${response.statusCode()}")
        }
        
        val uploadResponse = json.decodeFromString<BackgroundGeneratorUploadImageResponse>(response.body())
        
        if (uploadResponse.statusCode != 2000) {
            throw LightXBackgroundGeneratorException("Upload URL request failed: ${uploadResponse.message}")
        }
        
        return uploadResponse.body
    }
    
    private suspend fun uploadToS3(uploadUrl: String, imageFile: File, contentType: String) {
        val request = HttpRequest.newBuilder()
            .uri(URI.create(uploadUrl))
            .header("Content-Type", contentType)
            .PUT(HttpRequest.BodyPublishers.ofFile(imageFile.toPath()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXBackgroundGeneratorException("Image upload failed: ${response.statusCode()}")
        }
    }
    
    /**
     * Close the HTTP client
     */
    fun close() {
        httpClient.close()
    }
}

// MARK: - Exception Classes

class LightXBackgroundGeneratorException(message: String) : Exception(message)

// MARK: - Example Usage

object LightXBackgroundGeneratorExample {
    
    @JvmStatic
    suspend fun main(args: Array<String>) {
        try {
            // Initialize with your API key
            val lightx = LightXBackgroundGeneratorAPI("YOUR_API_KEY_HERE")
            
            val imageFile = File("path/to/input-image.jpg")
            
            if (!imageFile.exists()) {
                println("❌ Image file not found: ${imageFile.absolutePath}")
                return
            }
            
            // Get tips for better results
            lightx.getBackgroundTips()
            lightx.getBackgroundStyleSuggestions()
            lightx.getBackgroundPromptExamples()
            
            // Example 1: Generate background with professional style
            val professionalPrompts = lightx.getBackgroundStyleSuggestions()["professional"] ?: emptyList()
            val result1 = lightx.generateBackgroundWithPrompt(
                imageFile = imageFile,
                textPrompt = professionalPrompts[0],
                contentType = "image/jpeg"
            )
            println("🎉 Professional background result:")
            println("Order ID: ${result1.orderId}")
            println("Status: ${result1.status}")
            result1.output?.let { println("Output: $it") }
            
            // Example 2: Generate background with natural style
            val naturalPrompts = lightx.getBackgroundStyleSuggestions()["natural"] ?: emptyList()
            val result2 = lightx.generateBackgroundWithPrompt(
                imageFile = imageFile,
                textPrompt = naturalPrompts[0],
                contentType = "image/jpeg"
            )
            println("🎉 Natural background result:")
            println("Order ID: ${result2.orderId}")
            println("Status: ${result2.status}")
            result2.output?.let { println("Output: $it") }
            
            // Example 3: Generate backgrounds for different styles
            val styles = listOf("professional", "natural", "creative", "lifestyle", "outdoor")
            for (style in styles) {
                val prompts = lightx.getBackgroundStyleSuggestions()[style] ?: emptyList()
                val result = lightx.generateBackgroundWithPrompt(
                    imageFile = imageFile,
                    textPrompt = prompts[0],
                    contentType = "image/jpeg"
                )
                println("🎉 $style background result:")
                println("Order ID: ${result.orderId}")
                println("Status: ${result.status}")
                result.output?.let { println("Output: $it") }
            }
            
            // Example 4: Get image dimensions
            val (width, height) = lightx.getImageDimensions(imageFile)
            if (width > 0 && height > 0) {
                println("📏 Original image: ${width}x${height}")
            }
            
        } catch (e: Exception) {
            println("❌ Example failed: ${e.message}")
        }
    }
}

// MARK: - Coroutine Extension Functions

/**
 * Extension function to run the example in a coroutine scope
 */
fun runLightXBackgroundGeneratorExample() {
    runBlocking {
        LightXBackgroundGeneratorExample.main(emptyArray())
    }
}

File: 09-background-generator/LightXBackgroundGenerator.swift
/**
 * LightX AI Background Generator API Integration - Swift
 * 
 * This implementation provides a complete integration with LightX API v2
 * for AI-powered background generation functionality.
 */

import Foundation
import UIKit

// MARK: - Data Models

struct BackgroundGeneratorUploadImageResponse: Codable {
    let statusCode: Int
    let message: String
    let body: BackgroundGeneratorUploadImageBody
}

struct BackgroundGeneratorUploadImageBody: Codable {
    let uploadImage: String
    let imageUrl: String
    let size: Int
}

struct BackgroundGeneratorGenerationResponse: Codable {
    let statusCode: Int
    let message: String
    let body: BackgroundGeneratorGenerationBody
}

struct BackgroundGeneratorGenerationBody: Codable {
    let orderId: String
    let maxRetriesAllowed: Int
    let avgResponseTimeInSec: Int
    let status: String
}

struct BackgroundGeneratorOrderStatusResponse: Codable {
    let statusCode: Int
    let message: String
    let body: BackgroundGeneratorOrderStatusBody
}

struct BackgroundGeneratorOrderStatusBody: Codable {
    let orderId: String
    let status: String
    let output: String?
}

// MARK: - LightX Background Generator API Client

class LightXBackgroundGeneratorAPI {
    
    // MARK: - Properties
    
    private let apiKey: String
    private let baseURL = "https://api.lightxeditor.com/external/api"
    private let maxRetries = 5
    private let retryInterval: TimeInterval = 3.0 // seconds
    
    // MARK: - Initialization
    
    init(apiKey: String) {
        self.apiKey = apiKey
    }
    
    // MARK: - Public Methods
    
    /**
     * Upload image to LightX servers
     * - Parameters:
     *   - imageData: Image data to upload
     *   - contentType: MIME type (image/jpeg or image/png)
     * - Returns: Final image URL
     */
    func uploadImage(imageData: Data, contentType: String = "image/jpeg") async throws -> String {
        let fileSize = imageData.count
        
        guard fileSize <= 5242880 else { // 5MB limit
            throw LightXBackgroundGeneratorError.imageSizeExceeded
        }
        
        // Step 1: Get upload URL
        let uploadURL = try await getUploadURL(fileSize: fileSize, contentType: contentType)
        
        // Step 2: Upload image to S3
        try await uploadToS3(uploadURL: uploadURL, imageData: imageData, contentType: contentType)
        
        print("✅ Image uploaded successfully")
        return uploadURL.imageUrl
    }
    
    /**
     * Generate background
     * - Parameters:
     *   - imageURL: URL of the input image
     *   - textPrompt: Text prompt for background generation
     * - Returns: Order ID for tracking
     */
    func generateBackground(imageURL: String, textPrompt: String) async throws -> String {
        let endpoint = "\(baseURL)/v1/background-generator"
        
        guard let url = URL(string: endpoint) else {
            throw LightXBackgroundGeneratorError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = [
            "imageUrl": imageURL,
            "textPrompt": textPrompt
        ] as [String: Any]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXBackgroundGeneratorError.networkError
        }
        
        let backgroundResponse = try JSONDecoder().decode(BackgroundGeneratorGenerationResponse.self, from: data)
        
        guard backgroundResponse.statusCode == 2000 else {
            throw LightXBackgroundGeneratorError.apiError(backgroundResponse.message)
        }
        
        let orderInfo = backgroundResponse.body
        
        print("📋 Order created: \(orderInfo.orderId)")
        print("🔄 Max retries allowed: \(orderInfo.maxRetriesAllowed)")
        print("⏱️  Average response time: \(orderInfo.avgResponseTimeInSec) seconds")
        print("📊 Status: \(orderInfo.status)")
        print("💬 Text prompt: \"\(textPrompt)\"")
        
        return orderInfo.orderId
    }
    
    /**
     * Check order status
     * - Parameter orderID: Order ID to check
     * - Returns: Order status and results
     */
    func checkOrderStatus(orderID: String) async throws -> BackgroundGeneratorOrderStatusBody {
        let endpoint = "\(baseURL)/v1/order-status"
        
        guard let url = URL(string: endpoint) else {
            throw LightXBackgroundGeneratorError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = ["orderId": orderID]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXBackgroundGeneratorError.networkError
        }
        
        let statusResponse = try JSONDecoder().decode(BackgroundGeneratorOrderStatusResponse.self, from: data)
        
        guard statusResponse.statusCode == 2000 else {
            throw LightXBackgroundGeneratorError.apiError(statusResponse.message)
        }
        
        return statusResponse.body
    }
    
    /**
     * Wait for order completion with automatic retries
     * - Parameter orderID: Order ID to monitor
     * - Returns: Final result with output URL
     */
    func waitForCompletion(orderID: String) async throws -> BackgroundGeneratorOrderStatusBody {
        var attempts = 0
        
        while attempts < maxRetries {
            do {
                let status = try await checkOrderStatus(orderID: orderID)
                
                print("🔄 Attempt \(attempts + 1): Status - \(status.status)")
                
                switch status.status {
                case "active":
                    print("✅ Background generation completed successfully!")
                    if let output = status.output {
                        print("🖼️ Background image: \(output)")
                    }
                    return status
                    
                case "failed":
                    throw LightXBackgroundGeneratorError.processingFailed
                    
                case "init":
                    attempts += 1
                    if attempts < maxRetries {
                        print("⏳ Waiting \(retryInterval) seconds before next check...")
                        try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
                    }
                    
                default:
                    attempts += 1
                    if attempts < maxRetries {
                        try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
                    }
                }
                
            } catch {
                attempts += 1
                if attempts >= maxRetries {
                    throw error
                }
                print("⚠️  Error on attempt \(attempts), retrying...")
                try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
            }
        }
        
        throw LightXBackgroundGeneratorError.maxRetriesReached
    }
    
    /**
     * Complete workflow: Upload image and generate background
     * - Parameters:
     *   - imageData: Image data
     *   - textPrompt: Text prompt for background generation
     *   - contentType: MIME type
     * - Returns: Final result with output URL
     */
    func processBackgroundGeneration(imageData: Data, textPrompt: String, 
                                    contentType: String = "image/jpeg") async throws -> BackgroundGeneratorOrderStatusBody {
        print("🚀 Starting LightX AI Background Generator API workflow...")
        
        // Step 1: Upload image
        print("📤 Uploading image...")
        let imageURL = try await uploadImage(imageData: imageData, contentType: contentType)
        print("✅ Image uploaded: \(imageURL)")
        
        // Step 2: Generate background
        print("🖼️ Generating background...")
        let orderID = try await generateBackground(imageURL: imageURL, textPrompt: textPrompt)
        
        // Step 3: Wait for completion
        print("⏳ Waiting for processing to complete...")
        let result = try await waitForCompletion(orderID: orderID)
        
        return result
    }
    
    /**
     * Get background generation tips and best practices
     * - Returns: Dictionary of tips for better background results
     */
    func getBackgroundTips() -> [String: [String]] {
        let tips: [String: [String]] = [
            "input_image": [
                "Use clear, well-lit photos with good contrast",
                "Ensure the subject is clearly visible and centered",
                "Avoid cluttered or busy backgrounds",
                "Use high-resolution images for better results",
                "Good subject separation improves background generation"
            ],
            "text_prompts": [
                "Be specific about the background style you want",
                "Mention color schemes and mood preferences",
                "Include environmental details (indoor, outdoor, studio)",
                "Specify lighting preferences (natural, dramatic, soft)",
                "Keep prompts concise but descriptive"
            ],
            "general": [
                "Background generation works best with clear subjects",
                "Results may vary based on input image quality",
                "Text prompts guide the background generation process",
                "Allow 15-30 seconds for processing",
                "Experiment with different prompt styles for variety"
            ]
        ]
        
        print("💡 Background Generation Tips:")
        for (category, tipList) in tips {
            print("\(category): \(tipList)")
        }
        return tips
    }
    
    /**
     * Get background style suggestions
     * - Returns: Dictionary of background style suggestions
     */
    func getBackgroundStyleSuggestions() -> [String: [String]] {
        let styleSuggestions: [String: [String]] = [
            "professional": [
                "professional office background",
                "clean minimalist workspace",
                "modern corporate environment",
                "elegant business setting",
                "sophisticated professional space"
            ],
            "natural": [
                "natural outdoor landscape",
                "beautiful garden background",
                "scenic mountain view",
                "peaceful forest setting",
                "serene beach environment"
            ],
            "creative": [
                "artistic studio background",
                "creative workspace environment",
                "colorful abstract background",
                "modern art gallery setting",
                "inspiring creative space"
            ],
            "lifestyle": [
                "cozy home interior",
                "modern living room",
                "stylish bedroom setting",
                "contemporary kitchen",
                "elegant dining room"
            ],
            "outdoor": [
                "urban cityscape background",
                "park environment",
                "outdoor cafe setting",
                "street scene background",
                "architectural landmark view"
            ]
        ]
        
        print("💡 Background Style Suggestions:")
        for (category, suggestionList) in styleSuggestions {
            print("\(category): \(suggestionList)")
        }
        return styleSuggestions
    }
    
    /**
     * Get background prompt examples
     * - Returns: Dictionary of prompt examples
     */
    func getBackgroundPromptExamples() -> [String: [String]] {
        let promptExamples: [String: [String]] = [
            "professional": [
                "Modern office with glass windows and city view",
                "Clean white studio background with soft lighting",
                "Professional conference room with wooden furniture",
                "Contemporary workspace with minimalist design",
                "Elegant business environment with neutral colors"
            ],
            "natural": [
                "Beautiful sunset over mountains in the background",
                "Lush green garden with blooming flowers",
                "Peaceful lake with reflection of trees",
                "Golden hour lighting in a forest setting",
                "Serene beach with gentle waves and palm trees"
            ],
            "creative": [
                "Artistic studio with colorful paint splashes",
                "Modern gallery with white walls and track lighting",
                "Creative workspace with vintage furniture",
                "Abstract colorful background with geometric shapes",
                "Bohemian style room with eclectic decorations"
            ],
            "lifestyle": [
                "Cozy living room with warm lighting and books",
                "Modern kitchen with marble countertops",
                "Stylish bedroom with soft natural light",
                "Contemporary dining room with elegant table",
                "Comfortable home office with plants and books"
            ],
            "outdoor": [
                "Urban cityscape with modern skyscrapers",
                "Charming street with cafes and shops",
                "Park setting with trees and walking paths",
                "Historic architecture with classical columns",
                "Modern outdoor space with contemporary design"
            ]
        ]
        
        print("💡 Background Prompt Examples:")
        for (category, exampleList) in promptExamples {
            print("\(category): \(exampleList)")
        }
        return promptExamples
    }
    
    /**
     * Validate text prompt (utility function)
     * - Parameter textPrompt: Text prompt to validate
     * - Returns: Whether the prompt is valid
     */
    func validateTextPrompt(textPrompt: String) -> Bool {
        if textPrompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            print("❌ Text prompt cannot be empty")
            return false
        }
        
        if textPrompt.count > 500 {
            print("❌ Text prompt is too long (max 500 characters)")
            return false
        }
        
        print("✅ Text prompt is valid")
        return true
    }
    
    /**
     * Generate background with prompt validation
     * - Parameters:
     *   - imageData: Image data
     *   - textPrompt: Text prompt for background generation
     *   - contentType: MIME type
     * - Returns: Final result with output URL
     */
    func generateBackgroundWithPrompt(imageData: Data, textPrompt: String, 
                                     contentType: String = "image/jpeg") async throws -> BackgroundGeneratorOrderStatusBody {
        guard validateTextPrompt(textPrompt: textPrompt) else {
            throw LightXBackgroundGeneratorError.invalidPrompt
        }
        
        return try await processBackgroundGeneration(imageData: imageData, textPrompt: textPrompt, contentType: contentType)
    }
    
    /**
     * Get image dimensions (utility function)
     * - Parameter imageData: Image data
     * - Returns: (width, height) tuple
     */
    func getImageDimensions(imageData: Data) -> (Int, Int) {
        guard let image = UIImage(data: imageData) else {
            print("❌ Failed to create image from data")
            return (0, 0)
        }
        
        let width = Int(image.size.width)
        let height = Int(image.size.height)
        print("📏 Image dimensions: \(width)x\(height)")
        return (width, height)
    }
    
    // MARK: - Private Methods
    
    private func getUploadURL(fileSize: Int, contentType: String) async throws -> BackgroundGeneratorUploadImageBody {
        let endpoint = "\(baseURL)/v2/uploadImageUrl"
        
        guard let url = URL(string: endpoint) else {
            throw LightXBackgroundGeneratorError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = [
            "uploadType": "imageUrl",
            "size": fileSize,
            "contentType": contentType
        ] as [String: Any]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXBackgroundGeneratorError.networkError
        }
        
        let uploadResponse = try JSONDecoder().decode(BackgroundGeneratorUploadImageResponse.self, from: data)
        
        guard uploadResponse.statusCode == 2000 else {
            throw LightXBackgroundGeneratorError.apiError(uploadResponse.message)
        }
        
        return uploadResponse.body
    }
    
    private func uploadToS3(uploadURL: BackgroundGeneratorUploadImageBody, imageData: Data, contentType: String) async throws {
        guard let url = URL(string: uploadURL.uploadImage) else {
            throw LightXBackgroundGeneratorError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.httpBody = imageData
        
        let (_, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXBackgroundGeneratorError.uploadFailed
        }
    }
}

// MARK: - Error Handling

enum LightXBackgroundGeneratorError: Error, LocalizedError {
    case invalidURL
    case networkError
    case apiError(String)
    case imageSizeExceeded
    case uploadFailed
    case processingFailed
    case maxRetriesReached
    case invalidPrompt
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .networkError:
            return "Network error occurred"
        case .apiError(let message):
            return "API Error: \(message)"
        case .imageSizeExceeded:
            return "Image size exceeds 5MB limit"
        case .uploadFailed:
            return "Image upload failed"
        case .processingFailed:
            return "Background generation processing failed"
        case .maxRetriesReached:
            return "Maximum retry attempts reached"
        case .invalidPrompt:
            return "Invalid text prompt"
        }
    }
}

// MARK: - Example Usage

@MainActor
class LightXBackgroundGeneratorExample {
    
    static func runExample() async {
        do {
            // Initialize with your API key
            let lightx = LightXBackgroundGeneratorAPI(apiKey: "YOUR_API_KEY_HERE")
            
            // Get tips for better results
            lightx.getBackgroundTips()
            lightx.getBackgroundStyleSuggestions()
            lightx.getBackgroundPromptExamples()
            
            // Load image data (replace with your image loading logic)
            guard let inputImage = UIImage(named: "input_image"),
                  let inputImageData = inputImage.jpegData(compressionQuality: 0.8) else {
                print("❌ Failed to load input image")
                return
            }
            
            // Example 1: Generate background with professional style
            let professionalPrompts = lightx.getBackgroundStyleSuggestions()["professional"] ?? []
            let result1 = try await lightx.generateBackgroundWithPrompt(
                imageData: inputImageData,
                textPrompt: professionalPrompts[0],
                contentType: "image/jpeg"
            )
            print("🎉 Professional background result:")
            print("Order ID: \(result1.orderId)")
            print("Status: \(result1.status)")
            if let output = result1.output {
                print("Output: \(output)")
            }
            
            // Example 2: Generate background with natural style
            let naturalPrompts = lightx.getBackgroundStyleSuggestions()["natural"] ?? []
            let result2 = try await lightx.generateBackgroundWithPrompt(
                imageData: inputImageData,
                textPrompt: naturalPrompts[0],
                contentType: "image/jpeg"
            )
            print("🎉 Natural background result:")
            print("Order ID: \(result2.orderId)")
            print("Status: \(result2.status)")
            if let output = result2.output {
                print("Output: \(output)")
            }
            
            // Example 3: Generate backgrounds for different styles
            let styles = ["professional", "natural", "creative", "lifestyle", "outdoor"]
            for style in styles {
                let prompts = lightx.getBackgroundStyleSuggestions()[style] ?? []
                let result = try await lightx.generateBackgroundWithPrompt(
                    imageData: inputImageData,
                    textPrompt: prompts[0],
                    contentType: "image/jpeg"
                )
                print("🎉 \(style) background result:")
                print("Order ID: \(result.orderId)")
                print("Status: \(result.status)")
                if let output = result.output {
                    print("Output: \(output)")
                }
            }
            
            // Example 4: Get image dimensions
            let (width, height) = lightx.getImageDimensions(imageData: inputImageData)
            if width > 0 && height > 0 {
                print("📏 Original image: \(width)x\(height)")
            }
            
        } catch {
            print("❌ Example failed: \(error.localizedDescription)")
        }
    }
}
